src/librbd/migration/NBDStream.cc

```diff
@@ -0,0 +1,235 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/migration/NBDStream.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AsioEngine.h"
+#include "librbd/ImageCtx.h"
+#include <boost/asio/buffer.hpp>
+
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+
+#include <time.h>
+
+namespace librbd {
+namespace migration {
+
+namespace {
+
+const std::string URL_KEY  {"url"};
+const std::string PORT_KEY {"port"};
+
+} // anonymous namespace
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::NBDStream: " << this \
+                           << " " << __func__ << ": "
+
+template <typename I>
+struct NBDStream<I>::ReadRequest {
+  NBDStream*  nbd_stream;
+  io::Extents byte_extents;
+  bufferlist* data;
+  Context* on_finish;
+
+  ReadRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+              bufferlist* data, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      data(data), on_finish(on_finish) {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << dendl;
+  }
+
+  void send() {
+    data->clear();
+    read();
+  }
+
+  void read() {
+    auto cct = nbd_stream->m_cct;  
+    struct nbd_handle *nbd = nbd_stream->nbd;
+
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << dendl;
+      ldout(cct, 20) << "byte_length=" << byte_length << dendl;
+      auto ptr = buffer::ptr_node::create(buffer::create_small_page_aligned(
+        byte_length));
+      auto buffer = boost::asio::mutable_buffer(ptr->c_str(), byte_length);
+      data->push_back(std::move(ptr));
+      auto rc = nbd_aio_pread(nbd, boost::asio::buffer_cast<void *>(buffer),
+        byte_length, byte_offset, NBD_NULL_COMPLETION, 0);
+      if(rc == -1) {
+        lderr(cct) << "nbd_aio_pread: " << nbd_get_error() << 
+                        " (errno=" << nbd_get_errno() << ")" <<  dendl;
+        on_finish->complete(-EINVAL);
```

Comment:
nit: I would prefer complete with `-nbd_get_errno()` instead of `-EINVAL` here (and all other cases like this).